python/telemetry/delfipq.py

```diff
@@ -0,0 +1,415 @@
+    'Uptime' / BytesInteger(4),  # unit: s
+    'TotalUptime' / BytesInteger(4),
+    'TLMVersion' / BytesInteger(1),
+    'Telemetry' / Switch(this.TLMVersion, {
+        2: EPSTlmV2}),
+    'CellXmVoltage' / LinearAdapter(1000, BytesInteger(2))  # unit: V
+    )
+
+ADBTlm = Struct(
+    'Status' / SubsystemStatus,
+    'BootCounter' / BytesInteger(1),
+    'ResetCause' / ResetCause,
+    'Uptime' / BytesInteger(4),  # unit: s
+    'TotalUptime' / BytesInteger(4),
+    'TLMVersion' / BytesInteger(1),
+    'Telemetry' / Switch(this.TLMVersion, {
+        2: ADBTlmV2})
+    )
+
+ADBTlmFull = Struct(
+    'Status' / SubsystemStatus,
+    'BootCounter' / BytesInteger(1),
+    'ResetCause' / ResetCause,
+    'Uptime' / BytesInteger(4),  # unit: s
+    'TotalUptime' / BytesInteger(4),
+    'TLMVersion' / BytesInteger(1),
+    'Telemetry' / Switch(this.TLMVersion, {
+        2: ADBTlmV2}),
+    # Unit: degC
+    'Temperature' / LinearAdapter(10, BytesInteger(2, signed=True))
+    )
+
+COMMSTlm = Struct(
+    'Status' / SubsystemStatus,
+    'BootCounter' / BytesInteger(1),
+    'ResetCause' / ResetCause,
+    'Uptime' / BytesInteger(4),  # unit: s
+    'TotalUptime' / BytesInteger(4),
+    'TLMVersion' / BytesInteger(1),
+    'Telemetry' / Switch(this.TLMVersion, {
+        2: COMMSTlmV2})
+    )
+
+COMMSTlmFull = Struct(
+    'Status' / SubsystemStatus,
+    'BootCounter' / BytesInteger(1),
+    'ResetCause' / ResetCause,
+    'Uptime' / BytesInteger(4),  # unit: s
+    'TotalUptime' / BytesInteger(4),
+    'TLMVersion' / BytesInteger(1),
+    'Telemetry' / Switch(this.TLMVersion, {
+        2: COMMSTlmV2}),
+    # Unit: degC
+    'AmplifierTemperature' / LinearAdapter(10, BytesInteger(2, signed=True))
+    )
+
+PingMsg = Struct()
+
+Beacon = Struct(
+    'Destination' / Address,
+    'Size' / BytesInteger(1),
+    'BeaconSource' / Address,
+    'Service' / service,
+    'MessageType' / msgType,
+    'MessageOutcome' / msgOutcome,
+    'TLMSource' / Address,
+    'TelemetryHeader' / Switch(this.TLMSource, {
+        Address.OBC: OBCTlm,
+        Address.EPS: EPSTlm,
+        Address.ADB: ADBTlm,
+        Address.COMMS: COMMSTlm})
+    )
+
+DirectCommand = Struct(
+    'Service' / service,
+    'MessageType' / msgType,
+    'TelemetryHeader' / Switch(this.Service, {
+        service.Ping: PingMsg,
+        service.Telemetry: COMMSTlmFull})
+    )
+
+BusOverride = Struct(
+    'CommandID' / Address,
+    'Size' / BytesInteger(1),
+    'MessageSource' / Address,
+    'Service' / service,
+    'TLMSource' / Address,
+    'TelemetryHeader' / Switch(this.TLMSource, {
+        Address.OBC: OBCTlmFull,
+        Address.EPS: EPSTlmFull,
+        Address.ADB: ADBTlmFull,
+        Address.COMMS: COMMSTlm})
+    )
+
+BufferedCommand = Struct(
+    'CommandID' / Address,
+    'Destination' / Address,
+    'Size' / BytesInteger(1),
+    'Source' / Address,
+    'Service' / service,
+    'Request' / msgType,
+    'TelemetryHeader' / Switch(this.Service, {
```

Comment:
This one is incorrect. Tried to decode real-world message with buffered command, but failed. "service" type doesn't have OBC/EPS/ADB/COMMS values. Seems like typo?

Reply:
Interesting... buffered commands only get transmitted upon request so I guess you received a message sent when we were commanding. I did not add *all* messages Delfi-PQ can transmit because we never defined "a beacon message" but use the satellite as a network of systems, also exchanging internal debug commands (ping requests, for example, and some of those have not been properly tested before launch resulting in weird responses). We are working on a first sw patch to upload to the satellite to fix some of those inconsistencies...

Do you have the message hex?

Reply:
> ping requests, for example, and some of those have not been properly tested before launch resulting in weird responses

Hey! No issue at all. Just noticed this when I was adding support to r2cloud: https://github.com/dernasherbrezon/jradio/blob/master/src/main/java/ru/r2cloud/jradio/delfipq/DelfiPqBeacon.java 

Just wanted to say 'Thank you' for great delfipq tooling and documentation! 👍